infra/Makefile.toml

```diff
@@ -0,0 +1,597 @@
+extend = "cometbft-init"
+env = { "CMD" = "init", "NETWORK_NAME"="${NETWORK_NAME}", "CMT_DIR" = "${BASE_TESTNET_DIR}/node3/cometbft", "CMT_CONTAINER_NAME" = "cometbft-node3", "FLAGS" = "-a STDOUT -a STDERR --rm"  }
+
+[tasks.testnet-fendermint-config]
+dependencies = [
+  "testnet-config-newgenesis",
+  "testnet-config-peers",
+  "testnet-config-newkey",
+  "testnet-config-newaccount",
+  "testnet-config-newgateway",
+  "testnet-share-genesis"
+]
+
+[tasks.testnet-config-newgenesis]
+cwd = "./target/release"
+script = """
+./fendermint genesis --genesis-file $BASE_TESTNET_DIR/genesis.json \
+      new --chain-name testnet --base-fee ${BASE_FEE} --timestamp ${TIMESTAMP}
+
+cat $BASE_TESTNET_DIR/genesis.json
+"""
+
+[tasks.testnet-config-newkey]
+cwd = "./target/release"
+script="""
+mkdir -p $BASE_TESTNET_DIR/node0/fendermint/keys
+mkdir -p $BASE_TESTNET_DIR/node1/fendermint/keys
+mkdir -p $BASE_TESTNET_DIR/node2/fendermint/keys
+mkdir -p $BASE_TESTNET_DIR/node3/fendermint/keys
+
+./fendermint key gen --out-dir $BASE_TESTNET_DIR/node0/fendermint/keys --name validator0_key
+./fendermint key gen --out-dir $BASE_TESTNET_DIR/node1/fendermint/keys --name validator1_key
+./fendermint key gen --out-dir $BASE_TESTNET_DIR/node2/fendermint/keys --name validator2_key
+./fendermint key gen --out-dir $BASE_TESTNET_DIR/node3/fendermint/keys --name validator3_key
+
+"""
+
+[tasks.testnet-config-newaccount]
+cwd = "./target/release"
+script="""
+./fendermint genesis --genesis-file $BASE_TESTNET_DIR/genesis.json add-account \
+        --public-key $BASE_TESTNET_DIR/node0/fendermint/keys/validator0_key.pk --balance ${BALANCE}
+
+./fendermint genesis --genesis-file $BASE_TESTNET_DIR/genesis.json add-account \
+        --public-key $BASE_TESTNET_DIR/node1/fendermint/keys/validator1_key.pk --balance ${BALANCE}
+
+./fendermint genesis --genesis-file $BASE_TESTNET_DIR/genesis.json add-account \
+        --public-key $BASE_TESTNET_DIR/node2/fendermint/keys/validator2_key.pk --balance ${BALANCE}
+
+./fendermint genesis --genesis-file $BASE_TESTNET_DIR/genesis.json add-account \
+        --public-key $BASE_TESTNET_DIR/node3/fendermint/keys/validator3_key.pk --balance ${BALANCE}
+
+cat $BASE_TESTNET_DIR/genesis.json
+"""
+
+[tasks.testnet-config-peers]
+cwd = "./target/release"
+script="""
+./fendermint key add-peer --node-key-file $BASE_TESTNET_DIR/node0/cometbft/config/node_key.json --network-addr 192.167.10.2:26656 --local-peers-file $BASE_TESTNET_DIR/peers
+./fendermint key add-peer --node-key-file $BASE_TESTNET_DIR/node1/cometbft/config/node_key.json --network-addr 192.167.10.3:26656 --local-peers-file $BASE_TESTNET_DIR/peers
+./fendermint key add-peer --node-key-file $BASE_TESTNET_DIR/node2/cometbft/config/node_key.json --network-addr 192.167.10.4:26656 --local-peers-file $BASE_TESTNET_DIR/peers
+./fendermint key add-peer --node-key-file $BASE_TESTNET_DIR/node3/cometbft/config/node_key.json --network-addr 192.167.10.5:26656 --local-peers-file $BASE_TESTNET_DIR/peers
```

Comment:
Loop?

I am surprised a bit that all the p2p addresses use the same port. So these IPs work internally as well as externally? Then why are the p2p ports even mapped to the host?

Reply:
This is the same approach the cometbft testnet uses. Should we use a different approach?

Reply:
I think the CometBFT approach is good when you want something readable on the web, like in the docs you linked to. They are easy to read and doesn't require any extra knowledge like scripting, you just get the basics so there is no distraction. 

For our own infra scripts where there are not 4 but 12 containers, scripts we supposedly have to maintain and evolve, I would not stick verbatim to the CometBFT examples.

There is value in both, I know scripts can get confusing as well. I chose the scripting approach in the past so that I can create local nets of any number of nodes and for example automatically configure Prometheus to scrape all of them.

It's up to you :shrug:

Reply:
Sorry for confusion. I did not mean scripting. I meant this:

```
I am surprised a bit that all the p2p addresses use the same port. So these IPs work internally as well as externally? Then why are the p2p ports even mapped to the host?
```

Reply:
Okay, so the IPs work in the container, and they can be used with the internal ports. I am not sure why they should be mapped to the host, but why not, and I see CometBFT does the same :+1: